File: temas/abstraccion/ruby-01/src/robot.rb
class Robot
    attr_accessor :nombre
    attr_reader :material

    def initialize(nombre, material)
        @nombre = nombre
        @material = material
        @acciones = []
    end
        
    def mover(direccion)
        @acciones << "Moviendo hacia #{direccion}"
    end

    def girar(direccion)
        @acciones << "Girando hacia #{direccion}"
    end

    def ejecutar_acciones
        @acciones.each do |accion|
            puts accion
        end
    end

    def objetivo
        raise NotImplementedError, "Debe implementar el método objetivo en la subclase"
    end
end

File: temas/herencia/dart/lib/robot.dart
abstract class Robot {
  String nombre;
  String material;
  final List<String> acciones = [];

  Robot(this.nombre, this.material);

  //Acciones de cualquier robot

  void mover(String direccion) {
    acciones.add("$nombre moviéndose hacia $direccion");
  }

  void girar(String direccion) {
    acciones.add("$nombre girando hacia $direccion");
  }

  // Ejecutar las acciones del robot

  void ejecutarAcciones() {
    for (var accion in acciones) {
      print(accion);
    }
  }

  //Método abstracto
  String objetivo();
}
